src/poetry/utils/env/python_manager.py

```diff
@@ -0,0 +1,163 @@
+from __future__ import annotations
+
+import shutil
+import subprocess
+import sys
+
+from functools import cached_property
+from pathlib import Path
+from typing import TYPE_CHECKING
+
+from cleo.io.null_io import NullIO
+from cleo.io.outputs.output import Verbosity
+from poetry.core.constraints.version import Version
+from poetry.core.constraints.version import parse_constraint
+
+from poetry.utils._compat import decode
+from poetry.utils.env.exceptions import NoCompatiblePythonVersionFound
+from poetry.utils.env.script_strings import GET_PYTHON_VERSION_ONELINER
+
+
+if TYPE_CHECKING:
+    from cleo.io.io import IO
+
+    from poetry.config.config import Config
+    from poetry.poetry import Poetry
+
+
+class Python:
+    def __init__(self, executable: str | Path, version: Version | None = None) -> None:
+        self.executable = Path(executable)
+        self._version = version
+
+    @property
+    def version(self) -> Version:
+        if not self._version:
+            if self.executable == Path(sys.executable):
+                _python_version = ".".join(str(v) for v in sys.version_info[:3])
+            else:
+                encoding = "locale" if sys.version_info >= (3, 10) else None
+                _python_version = decode(
+                    subprocess.check_output(
+                        [str(self.executable), "-c", GET_PYTHON_VERSION_ONELINER],
+                        text=True,
+                        encoding=encoding,
+                    ).strip()
+                )
+            self._version = Version.parse(_python_version)
+
+        return self._version
+
+    @cached_property
+    def patch_version(self) -> Version:
+        return Version.from_parts(
+            major=self.version.major,
+            minor=self.version.minor,
+            patch=self.version.patch,
+        )
+
+    @cached_property
+    def minor_version(self) -> Version:
+        return Version.from_parts(major=self.version.major, minor=self.version.minor)
+
+    @staticmethod
+    def _full_python_path(python: str) -> Path | None:
+        # eg first find pythonXY.bat on windows.
+        path_python = shutil.which(python)
+        if path_python is None:
+            return None
```

Comment:
nit: 
```suggestion
        if path_python := shutil.which(python) is None:
            return None
```

Reply:
At work, we consider it bad style to use the walrus operator in an `if` statement even though the assigned variable is used outside of the `if` later. In other words, I do not like the suggested change.

Reply:
We have this is rule as well at work :smiley: